pkg/minikube/tunnel/tunnel.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2018 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tunnel
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/docker/machine/libmachine"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/minikube/pkg/minikube/config"
+)
+
+type tunnel interface {
+	cleanup() *Status
+	updateTunnelStatus() *Status
+}
+
+func errorTunnelAlreadyExists(id *ID) error {
+	return fmt.Errorf("there is already a running tunnel for this machine: %s", id)
+}
+
+func newTunnel(machineName string,
+	machineAPI libmachine.API,
+	configLoader config.Loader,
+	v1Core v1.CoreV1Interface, registry *persistentRegistry, router router) (*minikubeTunnel, error) {
+	clusterInspector := &clusterInspector{
+		machineName:  machineName,
+		machineAPI:   machineAPI,
+		configLoader: configLoader,
+	}
+	state, route, err := clusterInspector.getStateAndRoute()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine cluster info: %s", err)
+	}
+	id := ID{
+		Route:       route,
+		MachineName: machineName,
+		Pid:         getPid(),
+	}
+	runningTunnel, err := registry.IsAlreadyDefinedAndRunning(&id)
+	if err != nil {
+		return nil, fmt.Errorf("unable to check tunnel registry for conflict: %s", err)
+	}
+	if runningTunnel != nil {
+		return nil, fmt.Errorf("another tunnel is already running, shut it down first: %s", runningTunnel)
+	}
+
+	return &minikubeTunnel{
+		clusterInspector:     clusterInspector,
+		router:               router,
+		registry:             registry,
+		loadBalancerEmulator: newLoadBalancerEmulator(v1Core),
+		status: &Status{
+			TunnelID:      id,
+			MinikubeState: state,
+		},
+		reporter: &simpleReporter{
+			out: os.Stdout,
+		},
+	}, nil
+
+}
+
+type minikubeTunnel struct {
```

Comment:
It's difficult to tell the difference between the `minikubeTunnel` struct and `tunnel` struct. Can you adjust one of the names to be more distinct?

`minikube.tunnel.minikubeTunnel` is a rather perverse naming stutter, but since the struct name is private I won't worry too much about this.